benchmark/Parallel.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE CPP #-}
+-- |
+-- Module      : Main
+-- Copyright   : (c) 2018 Harendra Kumar
+--
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+
+import Control.DeepSeq (NFData)
+-- import Data.Functor.Identity (Identity, runIdentity)
+import System.Random (randomRIO)
+
+import Common (parseCLIOpts)
+
+import Streamly
+import Gauge
+
+import qualified Streamly.Benchmark.Prelude as Ops
+import qualified NestedOps as Nested
+
+{-# INLINE benchIONested #-}
+benchIONested :: (NFData b) => String -> (Int -> IO b) -> Benchmark
+benchIONested name f = bench name $ nfIO $ randomRIO (1,1) >>= f
+
+-- We need a monadic bind here to make sure that the function f does not get
+-- completely optimized out by the compiler in some cases.
+--
+-- | Takes a fold method, and uses it with a default source.
+{-# INLINE benchIO #-}
+benchIO :: (IsStream t, NFData b) => Int -> String -> (t IO Int -> IO b) -> Benchmark
+benchIO value name f = bench name $ nfIO $ randomRIO (1,1) >>= f . Ops.source value
+
+-- | Takes a source, and uses it with a default drain/fold method.
+{-# INLINE benchSrcIO #-}
+benchSrcIO
+    :: (t IO Int -> SerialT IO Int)
+    -> String
+    -> (Int -> t IO Int)
+    -> Benchmark
+benchSrcIO t name f
+    = bench name $ nfIO $ randomRIO (1,1) >>= Ops.toNull t . f
+
+{-# INLINE benchMonadicSrcIO #-}
+benchMonadicSrcIO :: String -> (Int -> IO ()) -> Benchmark
+benchMonadicSrcIO name f = bench name $ nfIO $ randomRIO (1,1) >>= f
```

Comment:
This and `benchIONested` are equivalent. You can remove this I guess?

Reply:
Fixed.